lightning/src/util/persist.rs

```diff
@@ -205,3 +213,760 @@ where
 	}
 	Ok(res)
 }
+
+struct MonitorUpdatingPersister<K: Deref, L: Deref>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	kv_store: K,
+	logger: L,
+}
+
+#[allow(dead_code)]
+impl<K: Deref, L: Deref> MonitorUpdatingPersister<K, L>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	pub fn read_channel_monitors_with_updates<
+		ES: Deref + Clone,
+		SP: Deref + Clone,
+		B: Deref,
+		F: Deref + Clone,
+	>(
+		&self,
+		entropy_source: ES,
+		signer_provider: SP,
+		broadcaster: &B,
+		fee_estimator: F,
+	) -> io::Result<Vec<(BlockHash,ChannelMonitor<<SP::Target as SignerProvider>::Signer>)>>
+	where
+		ES::Target: EntropySource + Sized,
+		SP::Target: SignerProvider + Sized,
+		B::Target: BroadcasterInterface,
+		F::Target: FeeEstimator,
+	{
+		let mut res = Vec::new();
+		// for each monitor...
+		for monitor_name in self.list_monitor_names()? {
+			// ...parse the monitor
+			let (bh, monitor) = self.deserialize_monitor(
+				entropy_source.clone(),
+				signer_provider.clone(),
+				&monitor_name,
+			)?;
+			// ...parse and apply the updates with an id higher than the monitor.
+			for update_name in self.list_update_names(&monitor_name)? {
+				let update = self.deserialize_monitor_update(&monitor_name, &update_name)?;
```

Comment:
Yeah, I debated this a bit through the iterations. The reason for doing this is:
  - Since there's as-you-go garbage collection, _most_ of the time you're going to be deserializing useful updates anyway
  - Trust-but-verify: I wanted to verify the update is the one the parsed `UpdateName` promises

But, as this has gotten a little more mature, I think we probably can just "trust" and not verify. I think you'll get an error if you try to apply these out of sequence, so we can just fail the entire read and get a human to sort out what went wrong.

Reply:
Done.